### PR TITLE
fix(router): rethrow exceptions

### LINKED ISSUE #2391: Router is not propagating exceptions on navigation
To reproduce, see https://github.com/jbdeboer/angular2_router_eating_stack_traces

This demo has a component which throws an exception in its constructor. When the component is created, that exception is not displayed in the Chrome DevTools console. It should be.

### PR DESCRIPTION
**Pull request by btford**
Closes #2391

### CODE REVIEW

**Comment by btford** [comment]
Assuming CI passes, this should be an easy merge.

**Comment by btford** [comment]
I made @matsko review this – he said LGTM. Merged as 5782f063f117be9d34c40d6ee4a3b8cf641b0d8d.

File: modules/angular2/test/router/router_integration_spec.ts [hunk 1]
```diff
@@ -47,6 +49,19 @@ export function main() {
              });
        }));
 
+    it('should rethrow exceptions from component constructors',
```
Comment (vicb) [inline modules/angular2/test/router/router_integration_spec.ts]: @btford does this test fail without the change in `router.ts` ?

File: modules/angular2/src/router/router.ts [hunk 2]
```diff
@@ -128,7 +128,10 @@ export class Router {
                 ObservableWrapper.callNext(this._subject, matchedInstruction.accumulatedUrl);
               });
 
-      PromiseWrapper.catchError(result, (_) => this._finishNavigating());
+      PromiseWrapper.catchError(result, (err) => {
+        this._finishNavigating();
+        return err;
```
Comment (vicb) [inline modules/angular2/src/router/router.ts]: noop ? (the return value is not used)

Comment (btford) [inline modules/angular2/src/router/router.ts]: This is the same as `.catch` – so the error will be propagated to anyone consuming this promise.

File: modules/angular2/test/router/router_integration_spec.ts (hunk 1 above)
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.ts]: yes

Comment (vicb) [inline modules/angular2/test/router/router_integration_spec.ts]: nope (tested this time)

File: modules/angular2/src/router/router.ts (hunk 2 above)
Comment (vicb) [inline modules/angular2/src/router/router.ts]: I think there are 2 anti-patterns here:

1) Promise should usually be chained

Current Code

```
commitP = this.commit(matchedInstruction);
returnedP = commitP.then(...);
ignoredP = returnedP.catchError(...);

return returnedP;
```

It should be
```
commitP = this.commit(matchedInstruction);
returnedP = commitP.then(thenFn);
returnedP = returnP.catchError(catchFn);

return returnedP;

// or or more concise

returnedP = PromiseWrapper.then(thenFn, catchFn);
```

Note than the concise way is a slightly different behavior: if an exception is thrown is the `thenFn`, the `catchFn` is not called. If you want to `catchFn` to also handle errors in the `thenFn`, you should go for the first code.

2) Behavior of `catchFn`

From an `Promise` reject function, you can either:
- Recover,
- Handle / Propagate the error

`p = rejectedP.catch(e => return value)` this is "recovery", p will be a `Promise` that resolves to `value` (it is not a *rejected* `Promise`.

`p = rejected.catch(e => throw new Error())` equivalent to  `p = rejected.catch(e => Promise.reject(new Error()))`, p will be a rejected `Promise` - the first form is probably more usual but sometimes [creates issue with TS](https://github.com/angular/angular/issues/2280) which expects a return value for the callback

Comment (btford) [inline modules/angular2/src/router/router.ts]: :+1: See https://github.com/angular/angular/pull/2565

**Pull request closed without merging**